src/consumer.ts

```diff
@@ -328,21 +333,47 @@ export class Consumer extends TypedEventEmitter {
   private async handleSqsResponse(
     response: ReceiveMessageCommandOutput,
   ): Promise<void> {
-    if (hasMessages(response)) {
-      if (this.handleMessageBatch) {
-        await this.processMessageBatch(response.Messages);
-      } else {
-        await Promise.all(
-          response.Messages.map((message: Message) =>
-            this.processMessage(message),
-          ),
-        );
-      }
-
-      this.emit("response_processed");
-    } else if (response) {
+    if (!hasMessages(response)) {
       this.emit("empty");
+      return;
     }
+
+    const messages = response.Messages;
+
+    if (this.handleMessageBatch) {
+      await this.processMessageBatch(messages);
+    } else {
+      let waitingMessages = 0;
+      await Promise.all(
+        messages.map(async (message) => {
+          while (
+            this.batchSize === 1 &&
+            this.inFlightMessages >= this.concurrency
+          ) {
+            if (waitingMessages === 0) {
+              this.emit("concurrency_limit_reached", {
+                limit: this.concurrency,
+                waiting: messages.length - this.inFlightMessages,
+              });
+            }
+            waitingMessages++;
+            await new Promise((resolve) =>
+              setTimeout(resolve, this.concurrencyWaitTimeMs),
```

Comment:
I need to think if `concurrencyWaitTimeMs`, it's probably too late and my brain is mush at this point though, but I think this conflicts with the polling timeout? Need to think on it.